core/object/message_queue.cpp

```diff
@@ -291,54 +264,208 @@ void MessageQueue::flush() {
-			read_pos += sizeof(Variant) * message->args;
+	pages_used = 1;
+	page_offset = 0;
+	page_messages[0] = 0;
+
+	mutex.unlock();
+}
+
+void CallQueue::statistics() {
+	mutex.lock();
+	HashMap<StringName, int> set_count;
+	HashMap<int, int> notify_count;
+	HashMap<Callable, int> call_count;
+	int null_count = 0;
+
+	for (uint32_t i = 0; i < pages_used; i++) {
+		uint32_t offset = 0;
+		for (uint32_t j = 0; j < page_messages[i]; j++) {
+			Page *page = pages[i];
+
+			//lock on each iteration, so a call can re-add itself to the message queue
+
+			Message *message = (Message *)&page->data[offset];
+
+			uint32_t advance = sizeof(Message);
+			if ((message->type & FLAG_MASK) != TYPE_NOTIFICATION) {
+				advance += sizeof(Variant) * message->args;
+			}
+
+			Object *target = message->callable.get_object();
+
+			if (target != nullptr) {
+				switch (message->type & FLAG_MASK) {
+					case TYPE_CALL: {
+						if (!call_count.has(message->callable)) {
+							call_count[message->callable] = 0;
+						}
+
+						call_count[message->callable]++;
+
+					} break;
+					case TYPE_NOTIFICATION: {
+						if (!notify_count.has(message->notification)) {
+							notify_count[message->notification] = 0;
+						}
+
+						notify_count[message->notification]++;
+
+					} break;
+					case TYPE_SET: {
+						StringName t = message->callable.get_method();
+						if (!set_count.has(t)) {
+							set_count[t] = 0;
+						}
+
+						set_count[t]++;
+
+					} break;
+				}
+
+			} else {
+				//object was deleted
+				print_line("Object was deleted while awaiting a callback");
+
+				null_count++;
+			}
+
+			//pre-advance so this function is reentrant
+			offset += advance;
+
+			if ((message->type & FLAG_MASK) != TYPE_NOTIFICATION) {
+				Variant *args = (Variant *)(message + 1);
+				for (int k = 0; k < message->args; k++) {
+					args[k].~Variant();
+				}
+			}
+
+			message->~Message();
 		}
 	}
 
+	print_line("TOTAL PAGES: " + itos(pages_used) + " (" + itos(pages_used * PAGE_SIZE_BYTES) + " bytes).");
+	print_line("NULL count: " + itos(null_count));
+
+	for (const KeyValue<StringName, int> &E : set_count) {
+		print_line("SET " + E.key + ": " + itos(E.value));
+	}
+
+	for (const KeyValue<Callable, int> &E : call_count) {
+		print_line("CALL " + E.key + ": " + itos(E.value));
+	}
+
+	for (const KeyValue<int, int> &E : notify_count) {
+		print_line("NOTIFY " + itos(E.key) + ": " + itos(E.value));
+	}
+
+	mutex.unlock();
+}
+
+bool CallQueue::is_flushing() const {
+	return flushing;
+}
+
+int CallQueue::get_max_buffer_usage() const {
+	return pages.size() * PAGE_SIZE_BYTES;
+}
+
+CallQueue::CallQueue(Allocator *p_custom_allocator, uint32_t p_max_pages, const String &p_error_text) {
+	if (p_custom_allocator) {
+		allocator = p_custom_allocator;
+		allocator_is_custom = true;
+	} else {
+		allocator = memnew(Allocator(16)); // 16 pages per allocator page, 64kb per allocator page. Anything small will do, though.
```

Comment:
```suggestion
		allocator = memnew(Allocator(16)); // 16 elements per allocator page, 64kb per allocator page. Anything small will do, though.
```

Reply:
16 PagedAllocator pages = 16 x 4 KiB = 64 KiB

The code could be a bit more explicit, like:
```
const uint32_t CallQueue::memnew(Allocator(TARGET_ALLOCATOR_PAGE_SIZE_BYTES = 64 * 1024;

// [...]

allocator = memnew(Allocator(TARGET_ALLOCATOR_PAGE_SIZE_BYTES / PAGE_SIZE_BYTES)); 
```